### PR TITLE
Feat/connectivity graph

### PR DESCRIPTION
**Pull request by ZenGround0**
This implements the connectivity graph described in issue #17.   It is marked as WIP because all tests are yet to be written.

This feature introduces a new `ctl` sub-command `ipfs-cluster-ctl peers graph `.   It triggers an RPC call that in turn triggers the `Peers` RPC call on all members tracked by the peer manager.  It then calls a new ipfs connector RPC endpoint `SwarmPeers` to gather info about all ipfs daemons' peers.  

After the graph data structure is passed back to the ctl process over the rest api, it is processed by dot-file building functionality to generate output.  Output defaults to stdin and is saved to a file on a flag.  

This has been run through the basics (minikube clusters between 2 - 4, with all nodes connected or one node disconnected), and everything works, including rendering the graph to an image by piping output to `dot`.

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

File: ipfs-cluster-ctl/graph.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
+package main
+
+import (
+	"errors"
+	"io"
+	"fmt"
+
+	"github.com/ipfs/ipfs-cluster/api"
+)
+
```
Comment (hsanjuan) [inline ipfs-cluster-ctl/graph.go]: I wouldn't mind a long comment here explaining how the dot file is built etc

File: ipfs-cluster-ctl/graph.go [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+	return nil
+}
+
+// truncate the provided peer id string to the 3 last characters.  Odds of 
+// pairwise collisions are less than 1 in 200,000 so with 70 cluster peers  
+// the chances of a collision are still less than 1 in 100 (birthday paradox).
+// As clusters grow bigger than this we can provide a flag for including
+// more characters.
+func shortId(peerString string) string {
```
Comment (hsanjuan) [inline ipfs-cluster-ctl/graph.go]: It probably make sense to use the same short peer identifiers as libp2p. when doing `String() on a peer it returns something like `Peer <abcdef>` where the original peer name was `Qmabcdef....`

File: ipfsconn/ipfshttp/ipfshttp.go [hunk 3]
```diff
@@ -92,6 +92,22 @@ type ipfsAddResp struct {
...
+type ipfsSwarmPeersResp struct {
+	Peers []ipfsPeer
+}
+
+type ipfsPeer struct {
+	Addr    string
+	Peer    string
+	Latency string
+	Muxer   string
+	Streams []ipfsStream
```
Comment (hsanjuan) [inline ipfsconn/ipfshttp/ipfshttp.go]: If you're not interested in the streams, you can just not include the key in your struct and the json parser will ignore it altogether.

Comment (hsanjuan) [reply ipfsconn/ipfshttp/ipfshttp.go]: (i remember posting here this too): Might work with just 

```
type ipfsSwarmPeersResp struct {
  Peers struct{
     Peer string
  }
}
}

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by ZenGround0** [comment]
Hey @hsanjuan this is ready for another pass of review.  All of your comments should be addressed, the code is now adequately tested, and the dot file writing has been factored out to go-dot.  There may be some trouble fetching the go-dot gx-dep on different machines because I am not sure I propagated the gx publish and my ipfs node comes on and offline with me.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: a4d522905f2560f059e232ea6f4c75f98544a5f2
  * Invalid signoff. Commit message must end with  
    **License: MIT  
Signed-off-by:**

Guidelines are available at https://github.com/ipfs/ipfs-cluster/blob/master/contribute.md

---
This message was auto-generated by https://gitcop.com

**Comment by hsanjuan** [comment]
@ZenGround0 Ok, I started reviewing from the oldest commit and made some comments, some of which are outright outdated by later commits (but check them though as they may not be).

You can use http://mars.i.ipfs.team:9444/ to pin your library in a way that is accessible.

File: ipfs-cluster-ctl/main.go [hunk 4]
```diff
@@ -225,6 +225,46 @@ cluster peers.
 						return nil
 					},
 				},
+				{
+					Name:  "graph",
```
[the remark this answers is not in the document]
Comment (ZenGround0) [reply ipfs-cluster-ctl/main.go]: Happy to change it back

File: ipfscluster_test.go [hunk 5]
```diff
@@ -1005,3 +1005,140 @@ func TestClustersRebalanceOnPeerDown(t *testing.T) {
 		t.Errorf("it should be pinned and is %s", s)
 	}
 }
+
+// Helper function for verifying cluster graph.  Will only pass if exactly the
+// peers in clusterIDs are fully connected to each other and the expected ipfs
+// mock connectivity exists.  Cluster peers not in clusterIDs are assumed to
+// be disconnected and the graph should reflect this
+func validateClusterGraph(t *testing.T, graph api.ConnectGraph, clusterIDs map[peer.ID]struct{}) {
```
[the remark this answers is not in the document]
Comment (ZenGround0) [reply ipfscluster_test.go]: I don't think so.  I tried to break my tests into units.  This function validates that the maps constructed by calling peers and discovering their neighbors is consistent with the expected cluster.  The graph_test file assumes a that these maps are correct and checks that the printing functions display them properly.

[the remark this answers is not in the document]
Comment (hsanjuan) [reply ipfscluster_test.go]: oki

**Comment by ZenGround0** [comment]
@hsanjuan, no hurry but I think this addresses your latest comments and is ready for another look

**Pull request merged**